### PR TITLE
Feat/add zeroex dex aggregator trades

### PR DESCRIPTION
**Pull request by bakabhai993**
This PR creates the deduped transactions table (ethereum only) for the 0x API + adds the zeroex data (ethereum only) in the dex aggregate table.

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [x] `coin_id` represents the ID of the coin on coinpaprika.com
* [x] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by bakabhai993** [comment]
Can someone be assigned to review this PR, @0xBoxer  ?

File: models/dex/aggregator/dex_aggregator_trades.sql [hunk 1]
```diff
@@ -8,34 +8,60 @@
...
+        ,maker_symbol           AS token_bought_symbol
+        ,taker_symbol           AS token_sold_symbol
+        ,token_pair
+        ,maker_token_amount     AS token_bought_amount
+        ,taker_token_amount     AS token_sold_amount
+        ,maker_token_amount_raw AS token_bought_amount_raw
+        ,taker_token_amount_raw AS token_sold_amount_raw
+        ,amount_usd
+        ,maker_token            AS token_bought_address
+        ,taker_token            AS token_sold_address
```
Comment (Hosuke) [inline models/dex/aggregator/dex_aggregator_trades.sql:53]: Hi @bakabhai993, do you think we can move the column renaming into the `zeroex_api_deduped_fills` model? So we can use the JINJA syntax to simplify the `dex_aggregator_trades` meta-table like `dex_trades` in the future.

Comment (bakabhai993) [reply models/dex/aggregator/dex_aggregator_trades.sql:53]: I am able to remove the renaming of the columns in this specific file, however, I would prefer to retain the current column names, specifically the **_zeroex_api_deduped_fills_** column. Would this be acceptable to you?

Comment (bakabhai993) [reply models/dex/aggregator/dex_aggregator_trades.sql:53]: I have already structured the zeroex_api_deduped_fills table on the lines of  dex_aggregator_trades FYI.

Comment (Hosuke) [reply models/dex/aggregator/dex_aggregator_trades.sql:53]: Please have a look to [this file](https://github.com/duneanalytics/spellbook/blob/main/models/dex/dex_trades.sql), if we move the renaming (`select xxx as xxx`) in to the sub-model, we may have the code in the meta-table clearer.
Moreover if `zeroex_api_deduped_fills` is a model need to merge into `dex_aggregator.trades` meta-table, its schema can follow the meta-table's schema(covers all columns of the meta-table but not limited to). If we want to retain the current column names in the `zeroex_api_deduped_fill`, we can have columns having different names but the same value in the schema.

eg.
```
SELECT
COALESCE(b.maker_symbol,b.maker_token) AS maker_symbol,
COALESCE(b.maker_symbol,b.maker_token) AS token_bought_symbol
```
In this way `SELECT token_bought_symbol FROM dex_model` still works.

Comment (Hosuke) [reply models/dex/aggregator/dex_aggregator_trades.sql:53]: Or we can have a wrapper model

Comment (bakabhai993) [reply models/dex/aggregator/dex_aggregator_trades.sql:53]: _**If we want to retain the current column names in the zeroex_api_deduped_fill, we can have columns having different names but the same value in the schema.**_

this should do the trick.

Comment (Hosuke) [reply models/dex/aggregator/dex_aggregator_trades.sql]: Done.

**Review by Hosuke — APPROVED** [review]
This PR adds `zeroex_api_ethereum_deduped_fills`, a derived model from `zeroex_api_ethereum_fills`, and the model is merged into `dex_aggregator_trades` meta-table.

I have reviewed this PR and here is the summary:

- Metadata & naming conventions look good ✅
- Config blocks match uniqueness tests in schema files ✅
- Tests and seeds added for the new models ✅
- Incremental filters added when joining large tables ✅

[Sample query here](https://dune.com/queries/1877969)

File: models/dex/aggregator/dex_aggregator_trades.sql (hunk 1 above)
Comment (bakabhai993) [reply models/dex/aggregator/dex_aggregator_trades.sql]: @RantumBits for your reference, we are adding extra columns in the deduped table for the reasons stated above.
@Hosuke thank you so much for the changes.

**Comment by jeff-dude** [comment]
@bakabhai993 please ignore CI tests failure for now. i can help fix that shortly. first, i'd like to share a few thoughts on how to reorg these models to help accomplish two things:
- keep 0x spells modeled in a way you & team want to see
- prepare for downstream consumption into sector-wide spells, i.e. `dex_aggregator.trades` -- this spell is early still and we'd like to get it off to a solid foundation

here are a few ideas i have to accomplish the above, please do let me know if you approve and then i can help push the changes:
- directory structure(s): `models/zeroex/<blockchain>` -- drop the sub-directories for nft/erc20, have all models within blockchain level. files names still allow for differentiation. we can then consolidate source/schema files as well
- find consistent approach for `zeroex` vs. `zeroex_api` usage. looks like decoded tables use schema name `zeroex_<blockchain>` and spells follow similar naming standard set in dbt_project yaml file, where `api` isn't used. looks like just file names and hardcoded value for project name leverage it. i'm thinking we drop `api` in file names, but can leave it in the column value for project name for `dex_aggregator.trades` if that is how we want it to be
- move any usage of `erc20` or `nft` into table names, rather than schema names, that way schema names remain clean similar to decoded tables for consistency
- leave your base-level 0x models as-is, but add new models on top which begin to standardize towards sector-wide spells -- that way, you don't have to repeat columns in your model just to help for downstream usage

i think that covers most of it -- happy to help push changes and revisit from there

**Comment by jeff-dude** [comment]
> @bakabhai993 please ignore CI tests failure for now. i can help fix that shortly. first, i'd like to share a few thoughts on how to reorg these models to help accomplish two things:
> 
> * keep 0x spells modeled in a way you & team want to see
> * prepare for downstream consumption into sector-wide spells, i.e. `dex_aggregator.trades` -- this spell is early still and we'd like to get it off to a solid foundation
> 
> here are a few ideas i have to accomplish the above, please do let me know if you approve and then i can help push the changes:
> 
> * directory structure(s): `models/zeroex/<blockchain>` -- drop the sub-directories for nft/erc20, have all models within blockchain level. files names still allow for differentiation. we can then consolidate source/schema files as well
> * find consistent approach for `zeroex` vs. `zeroex_api` usage. looks like decoded tables use schema name `zeroex_<blockchain>` and spells follow similar naming standard set in dbt_project yaml file, where `api` isn't used. looks like just file names and hardcoded value for project name leverage it. i'm thinking we drop `api` in file names, but can leave it in the column value for project name for `dex_aggregator.trades` if that is how we want it to be
> * move any usage of `erc20` or `nft` into table names, rather than schema names, that way schema names remain clean similar to decoded tables for consistency
> * leave your base-level 0x models as-is, but add new models on top which begin to standardize towards sector-wide spells -- that way, you don't have to repeat columns in your model just to help for downstream usage
> 
> i think that covers most of it -- happy to help push changes and revisit from there

@RantumBits just curious, can you help here too? since you've been pushing other 0x spells.

**Comment by RantumBits** [comment]
i updated this PR to remove updates to the dex_aggregator_trades table table for now

this PR now:
- adds new model for deduped table
- standardizes names 
- removes subdirectories 
- no updates to filenames (prefer in include _api ) 

note: this is for ethereum blockchain only. i'll update other blockchains in a new PR after this is complete.  

we'll submit a new PR with further updates to prep for downstream consumption & addition to aggregator_trades table

**Comment by jeff-dude** [comment]
thank you for helping finalize. i think we're on the same page now. here is quick summary of commits from my side:
- add dedupe seed to project file, similar to other two
- removed old directories nested in `zeroex/ethereum/` from project file to match your moving of files
- finalized schema/table names: `zeroex_ethereum.api_fills`, `zeroex_ethereum.api_fills_deduped`, `zeroex_ethereum.nft_fills`
- consolidated schema/source files (there were two of each in the `zeroex/ethereum/` directory, we only need one each, doesn't impact models themselves or naming above

**Comment by jeff-dude** [comment]
> thank you for helping finalize. i think we're on the same page now. here is quick summary of commits from my side:
> 
> * add dedupe seed to project file, similar to other two
> * removed old directories nested in `zeroex/ethereum/` from project file to match your moving of files
> * finalized schema/table names: `zeroex_ethereum.api_fills`, `zeroex_ethereum.api_fills_deduped`, `zeroex_ethereum.nft_fills`
> * consolidated schema/source files (there were two of each in the `zeroex/ethereum/` directory, we only need one each, doesn't impact models themselves or naming above

as you mentioned, next PR(s) can focus on same changes to each blockchain other than ethereum.

it would also be nice to see a follow-up PR to enhance the ethereum api fills model to match other blockchains we optimized recently (things like inner joins where possible, etc)

**Comment by RantumBits** [comment]
> > thank you for helping finalize. i think we're on the same page now. here is quick summary of commits from my side:
> > 
> > * add dedupe seed to project file, similar to other two
> > * removed old directories nested in `zeroex/ethereum/` from project file to match your moving of files
> > * finalized schema/table names: `zeroex_ethereum.api_fills`, `zeroex_ethereum.api_fills_deduped`, `zeroex_ethereum.nft_fills`
> > * consolidated schema/source files (there were two of each in the `zeroex/ethereum/` directory, we only need one each, doesn't impact models themselves or naming above
> 
> as you mentioned, next PR(s) can focus on same changes to each blockchain other than ethereum.
> 
> it would also be nice to see a follow-up PR to enhance the ethereum api fills model to match other blockchains we optimized recently (things like inner joins where possible, etc)

yes i will work on updating the model for a future PR

**Comment by RantumBits** [comment]
@jeff-dude i updated this PR with the change from PR 2534 https://github.com/duneanalytics/spellbook/pull/2534

**Comment by jeff-dude** [comment]
> @jeff-dude i updated this PR with the change from PR 2534 #2534

got it! let me get back into these PRs and move them forward

**Pull request merged**